File: src/PlantBasedPizza.Kitchen/application/BackgroundWorkers/Handlers/OrderConfirmedEventHandler.cs
using BackgroundWorkers.IntegrationEvents;
using PlantBasedPizza.Kitchen.Core.Adapters;
using PlantBasedPizza.Kitchen.Core.Entities;
using PlantBasedPizza.Kitchen.Core.Services;
using PlantBasedPizza.Shared.Guards;
using PlantBasedPizza.Shared.Logging;

namespace BackgroundWorkers.Handlers
{
    public class OrderConfirmedEventHandler
    {
        private readonly IKitchenEventPublisher _eventPublisher;
        private readonly IKitchenRequestRepository _kitchenRequestRepository;
        private readonly IRecipeService _recipeService;

        public OrderConfirmedEventHandler(IKitchenRequestRepository kitchenRequestRepository, IRecipeService recipeService, IKitchenEventPublisher eventPublisher)
        {
            _kitchenRequestRepository = kitchenRequestRepository;
            _recipeService = recipeService;
            _eventPublisher = eventPublisher;
        }
        
        public async Task Handle(OrderConfirmedEventV1 evt)
        {
            Guard.AgainstNull(evt, nameof(evt));

            var recipes = new List<RecipeAdapter>();

            foreach (var recipe in evt.Items)
            {
                recipes.Add(await this._recipeService.GetRecipe(recipe.RecipeIdentifier));
            }

            var kitchenRequest = new KitchenRequest(evt.OrderIdentifier, recipes);

            await this._kitchenRequestRepository.AddNew(kitchenRequest);
            await this._eventPublisher.PublishKitchenConfirmedOrderEventV1(kitchenRequest);
        }
    }
}

File: src/PlantBasedPizza.Kitchen.Node/src/commands/orderConfirmedEventHandler.ts
import { tracer } from "dd-trace";
import { RecipeAdapter } from "../adapters/recipeAdapter";
import { IKitchenEventPublisher } from "../entities/kitchenEventPublisher";
import { IKitchenRequestRepository } from "../entities/kitchenRepository";
import { KitchenRequest } from "../entities/kitchenRequest";
import { OrderConfirmedEvent } from "../integration-events/orderConfirmedEvent";
import { IRecipeService } from "../services/recipeService";

export class OrderConfirmedEventHandler {
  kitchenRequestRepository: IKitchenRequestRepository;
  eventPublisher: IKitchenEventPublisher;
  recipeService: IRecipeService;

  constructor(
    kitchenRequestRepository: IKitchenRequestRepository,
    eventPublisher: IKitchenEventPublisher,
    recipeService: IRecipeService,
  ) {
    this.kitchenRequestRepository = kitchenRequestRepository;
    this.eventPublisher = eventPublisher;
    this.recipeService = recipeService;
  }

  async handle(evt: OrderConfirmedEvent): Promise<void> {
    console.log(evt);

    const activeSpan = tracer.scope().active();
    activeSpan?.addTags({"order.orderIdentifier": evt.OrderIdentifier});
    activeSpan?.addTags({"order.inputItemCount": evt.Items.length});

    const existingOrder = await this.kitchenRequestRepository.retrieve(evt.OrderIdentifier);

    if (existingOrder !== null) {
      activeSpan?.addTags({"order.exists": "true"});
      return;
    }

    const recipes: RecipeAdapter[] = [];

    for (const item of evt.Items) {
      const recipe = await this.recipeService.getRecipe(item.RecipeIdentifier);

      if (recipe === undefined) {
        throw "Recipe not found";
      };

      recipes.push(recipe);
    }

    if (recipes.length != evt.Items.length) {
      const message = `Retrieved recipe count does not equal number of items on input: Retrieved ${recipes.length}. Input ${evt.Items.length}`;
      console.log(message);
      throw message;
    }

    activeSpan?.addTags({"order.recipesFound": recipes.length});

    const kitchenRequest = new KitchenRequest(evt.OrderIdentifier, recipes);

    await this.kitchenRequestRepository.addNew(kitchenRequest);
    await this.eventPublisher.publishKitchenOrderConfirmedEventV1({
        orderIdentifier: evt.OrderIdentifier
    });
  }
}
